smol/moca/sampler/kernel.py

```diff
@@ -408,6 +407,272 @@ def single_step(self, occupancy):
+            bin_size (float):
+                The enthalpy bin size to determine different states.
+            flatness (float): optional
+                The flatness factor used when checking histogram flatness.
+                Must be between 0 and 1.
+            mod_factor (float):
+                log of the initial modification factor used to update the
+                DOS/entropy.
+                Default is 1, which means the DOS shall be multiplied by e^1.
+            check_period (int): optional
+                The period in number of steps for the histogram flatness to be
+                checked.
+            update_period (int): optional
+                The period in number of steps to update the histogram and entropy.
+                In some cases periods slightly larger than each step, may allow
+                more efficient exploration of energy space.
+            mod_update (float|Callable): optional
+                A function used to update the fill factor when the histogram
+                satisfies the flatness criteria. It must monotonically decrease
+                to 0.
+                When a number is provided, at each update we will divide the
+                mod factor by this number.
+            seed (int): optional
+                non-negative integer to seed the PRNG
+            args:
+                Positional arguments to instantiate MCUsher.
+            kwargs:
+                Keyword arguments to instantiate MCUsher.
+        """
+        if min_enthalpy > max_enthalpy:
+            raise ValueError("min_enthalpy can not be larger than max_energy.")
+        if mod_factor <= 0:
+            raise ValueError("mod_factor must be greater than 0.")
+
+        self.flatness = flatness
+        self.check_period = check_period
+        self.update_period = update_period
+        self._m = mod_factor
+        self._window = (min_enthalpy, max_enthalpy, bin_size)
+        if callable(mod_update):
+            self._mod_update = mod_update
+        elif mod_update is not None:
+            self._mod_update = lambda x: x / mod_update
+        else:
+            self._mod_update = lambda x: x / 2.0
+
+        self._levels = np.arange(min_enthalpy, max_enthalpy, bin_size)
+
+        # The correct intialization will be handled in set_aux_state.
+        self._current_enthalpy = np.inf
+        self._current_features = np.zeros(len(ensemble.natural_parameters))
+        self._entropy = np.zeros(len(self._levels))
+        self._histogram = np.zeros(len(self._levels), dtype=int)
+        # Used to compute average features only.
+        self._occurrences = np.zeros(len(self._levels), dtype=int)
+        # Mean feature vector per level.
+        self._mean_features = np.zeros((len(self._levels),
+                                        len(ensemble.natural_parameters)))
+        self._steps_counter = 0  # Log n_steps elapsed.
+        self._states_counter = 0  # Log n_valid_states elapsed
```

Comment:
Yes, it is best to distinguish them. Since they are used for counts to be checked at different periods and can generally be reset at different times.

The `_steps_counter` is used to set when the histogram/entropy are updated (in certain cases it is more effective to not update those every single step (that's what the parameter `update_period` is for). This is also why I called it "update-counter" before. Can we rename  `_steps_counter`  to `_update_counter`?

The `_states_counter` is to know when to do a histogram flatness check, based on the value given for `check_period`. Similarly can we rename `_states_counter` to `_check_counter`?